### PR TITLE
Support CUDA nightly package in `tools/nightly.py`

### PR DESCRIPTION
**Pull request by XuehaiPan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131134
* __->__ #131133

Add a new option `--cuda` to `tools/nightly.py` to pull the nightly packages with CUDA support.

```bash
# installs pytorch-nightly with cpuonly
tools/nightly.py pull

# The following only available on Linux and Windows
# installs pytorch-nightly with latest CUDA we support
tools/nightly.py pull --cuda

# installs pytorch-nightly with CUDA 12.1
tools/nightly.py pull --cuda 12.1
```

Also add targets in `Makefile` and instructions in constribution guidelines.

```bash
# setup conda environment with pytorch-nightly
make setup-env

# setup conda environment with pytorch-nightly with CUDA support
make setup-env-cuda
```

### CODE REVIEW

File: tools/nightly.py [hunk 1]
```diff
@@ -412,23 +424,33 @@ def pull_nightly_version(spdir: str) -> None:
...
 
 def _get_listing_linux(source_dir: str) -> list[str]:
-    listing = glob.glob(os.path.join(source_dir, "*.so"))
-    listing.extend(glob.glob(os.path.join(source_dir, "lib", "*.so")))
-    return listing
+    return list(
+        itertools.chain(
+            glob.iglob(os.path.join(source_dir, "*.so")),
+            glob.iglob(os.path.join(source_dir, "lib", "*.so")),
+            glob.iglob(os.path.join(source_dir, "lib", "*.so.*")),
```
Comment (XuehaiPan) [inline tools/nightly.py:431]: Add glob `*.so.*` for CUDA libraries (e.g., `libcudnn.so.9`).

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**